src/bom/bitcode.rs

```diff
@@ -263,9 +277,62 @@ fn build_bitcode_arguments(chan : &mut mpsc::Sender<Option<Event>>,
 {
     let mut orig_target = None;
 
+    // Determine the language of the original input file, so that the bitcode can
+    // utilize the same language, because using C++ mode to compile C will cause
+    // errors:
+    //
+    // char* greeting() { return "hello, world"; }
+    //
+    // The above is perfectly acceptable C, but if compile as C++ it yields:
+    // error: ISO C++11 does not allow conversion from string literal to 'char*'.
+    let is_c_plusplus = orig_args.iter().find(
+        |&arg| arg.to_str().map(|a| !a.starts_with("-") &&
+                                (a.ends_with(".cc") ||
+                                 a.ends_with(".cpp")))
+            .unwrap_or_else(|| false))
+        .is_some();
+    let file_ext = if is_c_plusplus { ".cc" } else { ".c" };
+
+    // If the native compiler is used for preprocessing, it will write to a
+    // temporary output file that is subsequently consumed by clang to generate
+    // the bitcode.
+    let mut preprocess = ops.push_op(
+        &run(&Executable::new(orig_compiler_cmd,
+                              ExeFileSpec::Append,
+                              ExeFileSpec::option("-o")), &None)
+            .push_arg("-E")
+            .set_output_file(&FileArg::temp(file_ext)));
+
     let mut bcgen_op = ops.push_op(&run(&*CLANG_LLVM, bc_opts.clang_path)
                                    .set_label("clang:emit-llvm"));
 
+    if bc_opts.native_preproc {
+        // GCC and clang have internal directives (from system-level include
+        // files) for handling low-level things like alloc/dealloc resource
+        // management, atomic operations, etc.  If the native compiler is GCC,
+        // some of these directives will cause a failure on the clang bitcode
+        // generation path if they are still present.  Since the bitcode is being
+        // generated on a parallel effort and it is not being used for the final
+        // executable, these internal directives can be nullified to get valid
+        // bitcode with a minimal loss of information in that bitcode.
+        //
+        // If --strict is set, these are not applied and must be done manually
+        // via --inject-argument="-D..." if still needed.
+        if !bc_opts.strict {
+            bcgen_op.push_arg("-D__malloc__(X,Y)=");
+            bcgen_op.push_arg("-D__atomic_store(X,Y,Z)=");
+            bcgen_op.push_arg("-D__atomic_fetch_add(X,Y,Z)=0");
+            bcgen_op.push_arg("-D__atomic_fetch_sub(X,Y,Z)=0");
+            bcgen_op.push_arg("-D__atomic_fetch_and(X,Y,Z)=0");
+            bcgen_op.push_arg("-D__atomic_fetch_or(X,Y,Z)=0");
+            bcgen_op.push_arg("-D__atomic_compare_exchange(A,B,C,D,E,F)=0");
+            bcgen_op.push_arg("-D__atomic_exchange(A,B,C,D)=0");
+            bcgen_op.push_arg("-D__atomic_load(A,B,C)=0");
```

Comment:
This would be a good candidate to pull into a separate function too. I imagine this list is just going to grow.

Reply:
Done.